libs/ledger/src/consensus/consensus.cpp

```diff
@@ -278,17 +278,28 @@ bool Consensus::ValidBlockTiming(Block const &previous, Block const &proposed) c
     return false;
   }
 
-  assert(!qualified_cabinet_weighted.empty());
+  const uint64_t previous_block_window_ends = uint64_t(last_block_timestamp_ms + block_interval_ms_);
 
-  // First qual member can always produce
-  if (*qualified_cabinet_weighted.begin() == identity)
+  // Blocks cannot be created within the block interval of the previous, this enforces
+  // the block period
+  if (proposed_block_timestamp_ms < previous_block_window_ends)
   {
-    return true;
+    return false;
   }
 
-  // Until the time slot has elapsed, others can not produce
-  if (last_block_timestamp_ms + block_interval_ms_ < time_now_ms)
+  assert(!qualified_cabinet_weighted.empty());
+
+  // Miners must additionally wait N block periods according to their rank (0 being the best)
+  // Note, this is the identity specified in the block.
+  auto const miner_rank = std::distance(qualified_cabinet_weighted.begin(), std::find(qualified_cabinet_weighted.begin(), qualified_cabinet_weighted.end(), identity));
+
+  // Minting block. Timestamp: 5000 proposed: 1572486552000 Prev window ends: 528525664 last block TS:  1572486551000 miner rank: 12
+  if(proposed_block_timestamp_ms > uint64_t(previous_block_window_ends + (uint64_t(miner_rank) * block_interval_ms_)))
```

Comment:
This will need refactoring in future for the following:
-Non linear time slots for miner block production
-Possible moving average block time enforcement
-Time needs subsecond granularity